### PR TITLE
chore: Implement JSHandle for BiDi

### PR DESCRIPTION
**Pull request by Lightning00Blade**
Add Page.evaluateHandle when running BiDi protocol for Chrome or Firefox.
It covers only covers returning of JSHandle.

### CODE REVIEW

File: docs/api/puppeteer.jshandle.remoteobject.md [hunk 1]
```diff
@@ -4,16 +4,18 @@ sidebar_label: JSHandle.remoteObject
...
+  remoteObject():
+    | Protocol.Runtime.RemoteObject
+    | Bidi.CommonDataTypes.RemoteValue;
 }
 ```
 
 **Returns:**
 
-Protocol.Runtime.RemoteObject
+Protocol.Runtime.RemoteObject \| Bidi.CommonDataTypes.RemoteValue
```
Comment (OrKoN) [inline docs/api/puppeteer.jshandle.remoteobject.md]: Let's revert any changes to the API here.

File: packages/puppeteer-core/src/api/JSHandle.ts [hunk 2]
```diff
@@ -64,14 +67,14 @@ export class JSHandle<T = unknown> {
...
-  executionContext(): ExecutionContext {
+  executionContext(): ExecutionContext | Page {
     throw new Error('Not implemented');
   }
 
   /**
    * @internal
    */
-  get client(): CDPSession {
+  get client(): CDPSession | Connection {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/JSHandle.ts]: No changes should be necessary in this file too.

File: packages/puppeteer-core/src/common/bidi/Serializer.ts [hunk 3]
```diff
@@ -112,8 +126,25 @@ export class BidiSerializer {
...
 
-  static serialize(arg: unknown): Bidi.CommonDataTypes.LocalOrRemoteValue {
+  static serialize(
+    arg: unknown,
+    context: Page
+  ): Bidi.CommonDataTypes.LocalOrRemoteValue {
     // TODO: See use case of LazyArgs
+    const objectHandle = arg && arg instanceof JSHandle ? arg : null;
+    if (objectHandle) {
+      // TODO:
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/Serializer.ts]: what is this TODO about?

Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/bidi/Serializer.ts]: It was for the execution context match, at the time it was not implemented removing

File: packages/puppeteer-core/src/common/bidi/Page.ts [hunk 4]
```diff
@@ -67,8 +119,22 @@ export class Page extends PageBase {
...
+    return returnByValue
+      ? BidiSerializer.deserialize(result.result)
+      : getBidiHandler(this, result.result as Reference);
+  }
+}
+
+/**
+ * @internal
+ */
+export function getBidiHandler(context: Page, result: Reference): JSHandle {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/Page.ts]:
```suggestion
export function getBidiHandle(context: Page, result: Reference): JSHandle {
```

File: packages/puppeteer-core/src/common/bidi/JSHandle.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  override async getProperty<K extends keyof T>(
+    propertyName: HandleOr<K>
+  ): Promise<HandleFor<T[K]>> {
+    return await this.evaluateHandle((object, propertyName) => {
+      return object[propertyName as K];
+    }, propertyName);
+  }
+
+  override async getProperties(): Promise<Map<string, BaseJSHandle>> {
+    // TODO(lightning00blade): Temporary solution
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/JSHandle.ts]: could you describe what the final solution looks like?

Comment (OrKoN) [reply packages/puppeteer-core/src/common/bidi/JSHandle.ts]: (in the comment)

File: packages/puppeteer-core/src/common/bidi/JSHandle.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      return Object.getOwnPropertyNames(object);
+    });
+    const map: Map<string, BaseJSHandle> = new Map();
+    const results = await Promise.all(
+      keys.map(key => {
+        return this.getProperty(key);
+      })
+    );
+
+    keys.forEach((key, index) => {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/JSHandle.ts]: nit: let's use for (const [key, value] of Object.entries(keys))

File: packages/puppeteer-core/src/common/bidi/JSHandle.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+import {Page} from './Page.js';
+import {JSHandle as BaseJSHandle} from '../../api/JSHandle.js';
+import {BidiSerializer} from './Serializer.js';
+import {Connection} from './Connection.js';
+import * as Bidi from 'chromium-bidi/lib/cjs/protocol/protocol.js';
+
+export class JSHandle<T = unknown> extends BaseJSHandle<T> {
+  #disposed = false;
+  #context;
+  #remoteObject;
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/JSHandle.ts]: let's call it removeValue?

File: packages/puppeteer-core/src/common/bidi/utils.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import {Connection} from './Connection.js';
+
+/**
+ * @internal
+ */
+export const debugError = debug('puppeteer:error');
+/**
+ * @internal
+ */
+export async function releaseReference(
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/bidi/utils.ts:28]: it seems it would be suitable to be a method on the JSHandle instead of an util?

Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/bidi/utils.ts:28]: We will want to release the ConsoleLogEntry as the arguments for it are RemoteValue as well. I we do something similar for CDP.

File: packages/puppeteer-core/src/common/bidi/JSHandle.ts (hunk 7 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/common/bidi/JSHandle.ts]: `remoteValue`, typo

**Review by OrKoN — APPROVED** [review]
lgtm % suggestions

**Pull request merged**